src/mpid/ch4/src/ch4_part.c

```diff
@@ -14,24 +14,10 @@ int MPID_Psend_init(void *buf, int partitions, MPI_Aint count,
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PSEND_INIT);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PSEND_INIT);
 
-    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, dest);
-
     mpi_errno =
         MPIDIG_mpi_psend_init(buf, partitions, count, datatype, dest, tag, comm, info, request);
     MPIR_ERR_CHECK(mpi_errno);
 
-#ifndef MPIDI_CH4_DIRECT_NETMOD
-    if (MPIDI_REQUEST(*request, is_local)) {
-        mpi_errno = MPIDI_SHM_mpi_psend_init_hook(buf, partitions, count, datatype, dest, tag,
-                                                  comm, info, av, request);
-    } else
-#endif
-    {
-        mpi_errno = MPIDI_NM_mpi_psend_init_hook(buf, partitions, count, datatype, dest, tag,
-                                                 comm, info, av, request);
-    }
-    MPIR_ERR_CHECK(mpi_errno);
-
```

Comment:
OK, will do.

Reply:
Thinking about this some more, the receive side is still problematic. For precv_init, I want to fully initialize the request, then try to match it. If match exists, send handshake packet, otherwise defer until later. The complication lies in the scenario where the match occurs after MPI_START has been called. The generic layer doesn't know whether or not the lower layer is doing the actual communication, so it issues a CTS for the generic AM implementation when it might not be needed.

We could move this decision down to the lower layer, essentially adding `MPIDI_NM|SHM_precv_matched` APIs (and removing hooks I previously added). That way the generic CTS is only issued when desired.

Reply:
Why don't you implement a completely separate native handshake (using am but bypass the entire current MPIDIG code). That way, you can focus on the design for a good native implementation. Once you have it, we can look back and compare it with the current generic implementation to determine whether we can share some of the infrastructure. I think right now designing the hooks or figuring out the work around without the full picture is counter-productive.

Reply:
Yes, the hook vs. nm/shm API is probably not as important as the underlying logic. In the latest push I just moved the CTS out of the generic callback and into the nm/shm layer. I believe this should fully isolate the data transfer protocols. If I'm wrong, or there is no agreement, I will scrap it and copy/paste all the matching into the lower layer.